### PR TITLE
Azure cloud operation count metrics for azure cloud provider

### PR DESCRIPTION
**Pull request by kkmsft**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
Currently there are no metrics which records the number of cloud operations and also the ones which got rate limited. This PR adds two metrics - one for successful operation and other for rate limited operation.

This PR was arrived in the context of issue https://github.com/kubernetes/cloud-provider-azure/issues/228

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Testing performed by back porting to the cloud-provider-azure code.
TODO: Test after the new dependancy code is vendored in cloud-provider-azure.

**Does this PR introduce a user-facing change?**:

```release-note
Added cloud operation count metrics to azure cloud controller manager.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go [hunk 1]
```diff
@@ -201,7 +204,7 @@ func (az *azVirtualMachinesClient) CreateOrUpdate(ctx context.Context, resourceG
 		klog.V(10).Infof("azVirtualMachinesClient.CreateOrUpdate(%q, %q): end", resourceGroupName, VMName)
 	}()
 
-	mc := newMetricContext("vm", "create_or_update", resourceGroupName, az.client.SubscriptionID, source)
+	mc.Start()
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: why did you add this? I don't think this is required

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go [hunk 2]
```diff
@@ -191,7 +191,10 @@ func newAzVirtualMachinesClient(config *azClientConfig) *azVirtualMachinesClient
 
 func (az *azVirtualMachinesClient) CreateOrUpdate(ctx context.Context, resourceGroupName string, VMName string, parameters compute.VirtualMachine, source string) (resp *http.Response, err error) {
 	// /* Write rate limiting */
+	mc := newMetricContext("vm", "create_or_update", resourceGroupName, az.client.SubscriptionID, source)
+	mc.Count()
```
Comment (feiskyer) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: The count here is actually not true. When `mc.Count()` is invoked and ratelimited by the following line, then the requests won't be sent to Azure. 

I think we should move this before `az.client.CreateOrUpdate()`. Same all other clients.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go (hunk 1 above)
Comment (kkmsft) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: We want to measure the time taken for the specific cloud operation. This should not include time from any other items like logging, hence the need to start the timer.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go (hunk 2 above)
Comment (kkmsft) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: There are two metrics - one which is the count of operations  which is actually made by various components. And then the count which are ratelimited. We want to be able to know how many calls we are making and how many we are rate limiting. Just knowing how many were sent to azure is not enough information to arrive at conclusions. This gives us a better picture. Also, we can  gather the information of how many went to azure from these two counters.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go [hunk 3]
```diff
@@ -308,8 +320,11 @@ func newAzInterfacesClient(config *azClientConfig) *azInterfacesClient {
 }
 
 func (az *azInterfacesClient) CreateOrUpdate(ctx context.Context, resourceGroupName string, networkInterfaceName string, parameters network.Interface) (resp *http.Response, err error) {
+	mc := newMetricContext("interfaces", "create_or_update", resourceGroupName, az.client.SubscriptionID, "")
```
Comment (aramase) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: since the prefix and request in `newMetricContext` are mostly repeated, we should define it these as constants and use it.

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go (hunk 2 above)
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: I mean we should expose the counts sent to Azure as well as the ratelimited. The total count could be calculated by sum both.

**Review by feiskyer — COMMENTED** [review]
Thanks, KK. Could you squash the commits?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go (hunk 3 above)
Comment (khenidak) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: Let us have this change as part of this PR please. Let us have a private var, create on first call `NewMetricContext()`

Comment (kkmsft) [reply staging/src/k8s.io/legacy-cloud-providers/azure/azure_client.go]: As discussed, the metrics context is fresh for every operation.
